File: Higher Level/StackPot Ratio/app/Java/src/BetSizing.java

public class BetSizing {
	
	// print out sizing by street (eventually return an array)
	public static void returnSizes(Street c, double pct, int streetsRem){
		for(int i = 1; i <= streetsRem; i++){			
			Street prev = c;
			c = new Street(prev.returnFinalPot(), prev.returnFinalStack(), -1);
			c.setBetPct(pct);
			c.output();
		}
	}

	public static double testSize(Street c, double pct, int streetsRem){
		for(int i = 1; i < streetsRem; i++){			
			Street prev = c;
			c = new Street(prev.returnFinalPot(), prev.returnFinalStack(), -1);
			c.setBetPct(pct);
		}
		return c.error();
	}
	
	public static double binarySearch(double guess, double pot, double stack, int streets, double villbet){
		double mini = 0;
		double maxi = 10;
		Street prev = null;
		double err;
		double oldGuess = 0;
		double tempOld = 0;
		Street curr;
		
		while(true){
			curr = new Street(pot, stack, villbet);
			curr.setBetPct(guess);
			err = testSize(curr, guess, streets);
			
			if (Math.abs(err) < 0.0000001)
				return guess;
			else {
				if (prev == null) { // first iteration
					prev = curr;
					oldGuess = guess;
					if (err < 0) // bet size too big
						guess = guess/2; // halfway between 0 and curr
					else // bet size too small, halfway between 10 and curr
						guess = (maxi+guess)/2;
				}
				else{ // not first iteration
					if (guess < oldGuess){ // last attempt too big
						if (err < 0) {// still too big
							oldGuess = guess;
							guess = (guess+mini)/2; //halfway between min and prior guess
						}
						else { // was too big; now too small
							maxi = oldGuess; // the value has to be lower than this
							tempOld = guess;
							guess = (guess + oldGuess) / 2;
							oldGuess = tempOld;
						}
					}
					else { // last attempt too small
						if(err < 0){ // now too big
							mini = oldGuess;
							tempOld = guess;
							guess = (guess + oldGuess)/2;
							oldGuess = tempOld;
						}
						else{
							oldGuess = guess;
							guess = (maxi + guess)/2;
						}
					}
				}
			}
		}
	}
	
	public static void main(String []args){
		double pot = 10;
		double stack = 100;
		double villbet = 5;
		double guess = 0.5;
		int streets = 3;
		
		// start.output();		
		// returnSizes(start, pct, streetsRem, pct);
		// System.out.println(testSize(start, pct, streetsRem));
		double optimal = binarySearch(guess, pot, stack, streets, villbet);
		returnSizes(new Street(pot, stack, villbet), optimal, streets);
	
		/* TODO:
		 * - simple web app: takes the 5 inputs in a form, return bet %
		 * - gradually build from there
		 */
	}
}

File: Higher Level/StackPot Ratio/app/Bet Sizing.py
import numpy as np
import math
import pandas as pd
from pandas import *

def next_street_bet(pot, stack, streets_remaining, villbet, betpct):
    if villbet < 0: # it's not the first street
        bet = pot*betpct
    else:
        bet = (villbet*2+pot)*betpct+villbet
    if streets_remaining == 1:
        if abs(stack-bet) < 0.000001:
            return True,betpct
        else:
            return False,stack-bet
    else:
        return next_street_bet(pot+bet*2, stack-bet, streets_remaining-1, -1, betpct)

def binary_search(curr, params):    
    mini = 0
    maxi = 10
    prev = float('nan')
    while True:
        bet_size = next_street_bet(params[0],params[1],params[2],params[3],curr)
            
        if bet_size[0] == True: # found it
            return bet_size[1]
        else: # keep going
            if math.isnan(prev): # first iteration
                prev = curr
                if bet_size[1] < 0: # bet size too big
                    curr /= 2 # halfway between 0 and curr
                else:
                    curr = (1+curr)/2 # halfway between 1 and curr
            else:
                if curr < prev: # if last attempt was too big
                    if bet_size[1] < 0: # still too big
                        prev = curr
                        curr = (curr+mini)/2
                    else: # now too small
                        maxi = prev # this is the highest possible value
                        tempprev = curr
                        curr = (curr + prev)/2
                        prev = tempprev
                else: # if last attempt was too small
                    if bet_size[1] < 0: # now too big
                        mini = prev # this is the lowest possible value
                        tempprev = curr
                        curr = (curr + prev)/2
                        prev = tempprev
                    else: # still too small
                        prev = curr
                        curr = (maxi+curr)/2
    return False # didn't work somehow

# bet needs: pot, stack, streets_remaining
print binary_search(0.5, [40, 100, 3, -1])
# raise needs: pot, stack, betraisepct, streets_remaining, vill_bet
print binary_search(0.5, [10, 100, 2, 5])

def get_sizes(pot, stack, betpct, streets_remaining, villbet, sizes):
    
    if villbet == -1: # not the first street
        bet = betpct*pot
        sizes.append([pot,stack,float('nan'),bet,betpct])

    else:
        bet = (villbet*2 + pot)*betpct + villbet
        sizes.append([pot,stack,villbet,bet,betpct])

    if streets_remaining == 1: # base case
        return sizes
    
    return get_sizes(pot + 2*bet, stack-bet, betpct, streets_remaining-1, -1, sizes)
    

pot = 40
villbet = -1 # -1 indicates it is checked to Hero
stack = 100
streets_remaining = 3
bet_sizes = get_sizes(pot, stack, \
            binary_search(0.5, [pot, stack, streets_remaining, villbet]),\
            streets_remaining, villbet, [])

df = DataFrame(bet_sizes)
df.columns = 'Pot','Stack','Bet/Raise','Villain Bet','%'
df.index.name = 'Street #'
print df


pot = 10
villbet = 5
stack = 100
streets_remaining = 3
bet_sizes = get_sizes(pot, stack, \
            binary_search(0.5, [pot, stack, streets_remaining, villbet]),\
            streets_remaining, villbet, [])

df = DataFrame(bet_sizes)
df.columns = 'Pot','Stack','Bet/Raise','Villain Bet','%'
df.index.name = 'Street #'
print df


